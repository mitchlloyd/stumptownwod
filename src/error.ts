export const networkErrorMessage = 'There was a network error when visiting the website.';

export function statusErrorMessage(statusCode: number) {
  return `The server returned an error: ${statusCode}`;
}

export const readBodyErrorMessage = 'There was an error reading the body of the request.';
