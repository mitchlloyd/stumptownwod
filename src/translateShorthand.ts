export function translateShorthand(text: string) {
  const tokens = text.split(/\s/);

  return tokens.map(token => {
    return process(
      token,
      [
        nTimes,
        chestToBar,
        handStandPushUps,
        timeDuration,
        calories,
        hyphenatedRepScheme,
      ],
    );
  }).join(' ');
}

interface StringProcessor {
  (text: string): string;
}

function process(value: string, fns: StringProcessor[]) {
  let result: string = value;

  fns.forEach(fn => result = fn(result));

  return result;
}

function nTimes(text: string) {
  return text.replace(/^x(\d+)$/i, '$1 times');
}

function chestToBar(text: string) {
  return text.replace(/^c2b$/i, 'chest to bar');
}

function handStandPushUps(text: string) {
  return text.replace(/^hspu$/i, 'hand stand push-ups');
}

function timeDuration(text: string) {
  return text.replace(/^(\d+):(\d+)$/i, `<say-as interpret-as="time">$1'$2"</say-as>`)
}

function calories(text: string) {
  return text.replace(/^cals$/i, 'calories')
}

function hyphenatedRepScheme(text: string) {
  if (/^\d+(-\d+)+$/.test(text)) {
    const nums = text.split('-');
    const lastNum = nums.pop();

    return `sets of ${nums.join(', ')}, and ${lastNum}`;
  }

  return text;
}
