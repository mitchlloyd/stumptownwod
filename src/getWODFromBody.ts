import * as cheerio from 'cheerio';
import { translateShorthand } from './translateShorthand';

const WARM_UP = /(warm[- ]up|therapydia)/i;
const ONLY_WHITESPACE = /^\s+$/;

export function getWODFromBody(body: string) {
  const doc = cheerio.load(body);

  const postContent = doc('h2:contains("CrossFit Stumptown")').parent();
  const metcons = postContent.find('h3').not((_, element) => {
    return WARM_UP.test(cheerio(element).text());
  });

  return metcons.map((_, el) => {
    return parseMetcon(cheerio(el));
  }).get().join('\n');
}

function parseMetcon(heading: Cheerio) {
  const headingMessage = `<p>${heading.text()}</p>`;

  const instructionSet = heading.next('p').contents().map((_, element) => {
    if (element.type === 'text' && !ONLY_WHITESPACE.test(element.data)) {
      return element
    }
  });

  const instructionElements: CheerioElement[] = instructionSet.get() as any[];

  const instructions = instructionElements.map((el) => {
    return `<s>${translateShorthand(el.data)}</s>`
  });

  return [headingMessage, ...instructions].join('\n');
}
