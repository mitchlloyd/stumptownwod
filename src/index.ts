import * as Alexa from 'alexa-sdk';
import fetch from 'node-fetch';
import * as Error from './error';
import { getWODFromBody } from './getWODFromBody';

const handlers: Alexa.Handlers<{}> = {
  'LaunchRequest'() {
    this.emit('GetWODForToday');
  },

  async 'GetWODForToday'() {
    const wod = await getWOD();
    this.emit(':tell', wod);
  }
};

interface LambdaHandler<T> {
  (event: Alexa.RequestBody<T>, context: Alexa.Context, callback?: (err: any, response: any) => void): void;
}

export const handler: LambdaHandler<{}> = (event, context, callback) => {
  const alexa = Alexa.handler(event, context, callback);
  alexa.registerHandlers(handlers);
  alexa.execute();
};

async function getWOD(): Promise<string> {
  let response;

  try {
    response = await fetch('http://www.crossfitstumptown.com/wod/');
  } catch (e) {
    return Error.networkErrorMessage;
  }

  if (!response.ok) {
    return Error.statusErrorMessage(response.status);
  }

  let body;

  try {
    body = await response.text();
  } catch (e) {
    return Error.readBodyErrorMessage;
  }

  return getWODFromBody(body);
}

// End to end test for today:
// getWOD().then(console.log)
