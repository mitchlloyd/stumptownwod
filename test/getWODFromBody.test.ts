import { getWODFromBody } from '../src/getWODFromBody';
import { loadFixture } from './utils/loadFixture';

test('responding for a standard workout', () => {
  const html = loadFixture('standard');

  expect(getWODFromBody(html)).toMatchSnapshot();
});

test('responding for a hero workout', () => {
  const html = loadFixture('hero');

  expect(getWODFromBody(html)).toMatchSnapshot();
});

test('responding for a 2-metcon workout', () => {
  const html = loadFixture('2-metcon');

  expect(getWODFromBody(html)).toMatchSnapshot();
});

test('responding for a hatch-squat workout', () => {
  const html = loadFixture('hatch-squat');

  expect(getWODFromBody(html)).toMatchSnapshot();
});
