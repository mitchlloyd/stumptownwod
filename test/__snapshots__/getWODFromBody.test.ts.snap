// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responding for a 2-metcon workout 1`] = `
"<p>Metcon (3 Rounds for calories)</p>
<s>12 Goblet squats 70/53</s>
<s><say-as interpret-as=\\"time\\">2'00\\"</say-as> Max calories on Bike</s>
<s>Rest: 2 minutes</s>
<s>3 times</s>
<p>Metcon (3 Rounds for distance)</p>
<s>12 Clean High Pulls</s>
<s><say-as interpret-as=\\"time\\">2'00\\"</say-as> Max Meters on Rower</s>
<s>Rest 2 minutes</s>
<s>3 times</s>"
`;

exports[`responding for a hatch-squat workout 1`] = `
"<p>Back Squat (Hatch: Week 9/ Day 2)</p>
<p>Front Squat (Hatch: Week 9/ Day 2)</p>
<p>Metcon (Time)</p>
<s>sets of 21, 15, and 9</s>
<s>chest to bar Pull-ups</s>
<s>hand stand push-ups</s>"
`;

exports[`responding for a hero workout 1`] = `
"<p>Coffland (Time)</p>
<s>Hang from a pull-up bar for 6 minutes</s>
<s>Each time you drop from the bar, perform:</s>
<s>800-m run</s>
<s>30 push-ups</s>"
`;

exports[`responding for a standard workout 1`] = `
"<p>Metcon (Time)</p>
<s>For Time:</s>
<s>1 Mile Run</s>
<s>100 Box Jumps 24/20</s>
<s>800m Run</s>
<s>100 Wall Balls 20/14</s>
<s>1 Mile Run</s>"
`;
